pkg/handlers/docker/mutation/customimage/inject.go

```diff
@@ -0,0 +1,215 @@
+	"context"
+	_ "embed"
+	"fmt"
+	"strings"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/exp/runtime/topologymutation"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/d2iq-labs/capi-runtime-extensions/api/v1alpha1"
+	commonhandlers "github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/handlers"
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/handlers/mutation"
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/patches"
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/patches/selectors"
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/variables"
+	"github.com/d2iq-labs/capi-runtime-extensions/pkg/handlers/generic/clusterconfig"
+)
+
+const (
+	// HandlerNamePatch is the name of the inject handler.
+	HandlerNamePatch = "DockerCustomImagePatch"
+
+	defaultKinDImageRepository = "ghcr.io/mesosphere/kind-node"
+)
+
+type customImagePatchHandler struct {
+	variableName      string
+	variableFieldPath []string
+}
+
+var (
+	_ commonhandlers.Named     = &customImagePatchHandler{}
+	_ mutation.GeneratePatches = &customImagePatchHandler{}
+	_ mutation.MetaMutater     = &customImagePatchHandler{}
+)
+
+func NewPatch() *customImagePatchHandler {
+	return newCustomImagePatchHandler(variableName)
+}
+
+func NewMetaPatch() *customImagePatchHandler {
+	return newCustomImagePatchHandler(clusterconfig.MetaVariableName, variableName)
+}
+
+func newCustomImagePatchHandler(
+	variableName string,
+	variableFieldPath ...string,
+) *customImagePatchHandler {
+	return &customImagePatchHandler{
+		variableName:      variableName,
+		variableFieldPath: variableFieldPath,
+	}
+}
+
+func (h *customImagePatchHandler) Name() string {
+	return HandlerNamePatch
+}
+
+func (h *customImagePatchHandler) Mutate(
+	ctx context.Context,
+	obj runtime.Object,
+	vars map[string]apiextensionsv1.JSON,
+	holderRef runtimehooksv1.HolderReference,
+	_ client.ObjectKey,
+) error {
+	log := ctrl.LoggerFrom(ctx).WithValues(
+		"holderRef", holderRef,
+	)
+
+	customImageVar, found, err := variables.Get[v1alpha1.OCIImage](
+		vars,
+		h.variableName,
+		h.variableFieldPath...,
+	)
+	if err != nil {
+		return err
+	}
+	if !found {
+		log.V(5).Info("Docker customImage variable not defined, using default KinD node image")
+	}
+
+	log = log.WithValues(
+		"variableName",
+		h.variableName,
+		"variableFieldPath",
+		h.variableFieldPath,
+		"variableValue",
+		customImageVar,
+	)
+
+	err = patches.Generate(
+		obj,
+		vars,
+		&holderRef,
+		selectors.InfrastructureWorkerMachineTemplates("v1beta1", "DockerMachineTemplate"),
+		log,
+		func(obj *unstructured.Unstructured) error {
```

Comment:
By virtue of being applied on docker cluster class, yes I think so, but I agree there should be a check here just in case it gets applied to a wrong cluster provider.

Reply:
Actually, yes the selector can only select objects of kind `DockerMachineTemplate` in the `infrastructure.cluster.x-k8s.io/v1beta1` API version so this is safe